recipes/IWSLT22_lowresource/train.py

```diff
@@ -0,0 +1,400 @@
+    def sp_audio_pipeline(wav):
+        """Load the audio signal. This is done on the CPU in the `collate_fn`."""
+        sig = sb.dataio.dataio.read_audio(wav)
+        sig = sig.unsqueeze(0)
+        sig = hparams["speed_perturb"](sig)
+        sig = sig.squeeze(0)
+        return sig
+
+    # Define text processing pipeline. We start from the raw text and then
+    # encode it using the tokenizer. The tokens with BOS are used for feeding
+    # decoder during training, the tokens with EOS for computing the cost function.
+    @sb.utils.data_pipeline.takes("trans")
+    @sb.utils.data_pipeline.provides(
+        "trans", "tokens_list", "tokens_bos", "tokens_eos",
+    )
+    def reference_text_pipeline(translation):
+        """Processes the transcriptions to generate proper labels"""
+        yield translation
+        tokens_list = hparams["tokenizer"].encode_as_ids(translation)
+        yield tokens_list
+        tokens_bos = torch.LongTensor([hparams["bos_index"]] + (tokens_list))
+        yield tokens_bos
+        tokens_eos = torch.LongTensor(tokens_list + [hparams["eos_index"]])
+        yield tokens_eos
+
+    datasets = {}
+    data_folder = hparams["data_folder"]
+    for dataset in ["train", "valid"]:
+        json_path = f"{data_folder}/{dataset}.json"
+
+        is_use_sp = dataset == "train" and "speed_perturb" in hparams
+        audio_pipeline_func = sp_audio_pipeline if is_use_sp else audio_pipeline
+
+        datasets[dataset] = sb.dataio.dataset.DynamicItemDataset.from_json(
+            json_path=json_path,
+            replacements={"data_root": data_folder},
+            dynamic_items=[audio_pipeline_func, reference_text_pipeline],
+            output_keys=[
+                "id",
+                "sig",
+                "duration",
+                "trans",
+                "tokens_list",
+                "tokens_bos",
+                "tokens_eos",
+            ],
+        )
+
+    for dataset in ["valid", "test"]:
+        json_path = f"{data_folder}/{dataset}.json"
+        datasets[dataset] = sb.dataio.dataset.DynamicItemDataset.from_json(
+            json_path=json_path,
+            replacements={"data_root": data_folder},
+            dynamic_items=[audio_pipeline, reference_text_pipeline],
+            output_keys=[
+                "id",
+                "sig",
+                "duration",
+                "trans",
+                "tokens_list",
+                "tokens_bos",
+                "tokens_eos",
+            ],
+        )
+
+    # Sorting training data with ascending order makes the code  much
+    # faster  because we minimize zero-padding. In most of the cases, this
+    # does not harm the performance.
+    if hparams["sorting"] == "ascending":
+        # use smaller dataset to debug the model
+        if hparams["debug"]:
+            datasets["train"] = datasets["train"].filtered_sorted(
+                key_min_value={"duration": 1},
+                key_max_value={"duration": 5},
```

Comment:
We should avoid have values hardcoded. COuld 1 and 5 be put in the yaml as hyperparams instead?